### ISSUE TITLE
no-shadow false positive in double script block

### ISSUE THREAD

**Issue by sqal** [vuejs/eslint-plugin-vue#1800]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 8.9.0
- **eslint-plugin-vue version:** 8.4.1
- **Node version:** 
- **Operating System:**  

**Please show your full configuration:**

```json5
module.exports = {
  extends: [
    'plugin:vue/vue3-recommended',
  ],
  rules: {
    'no-shadow': 'error',
  },
};
```

**What did you do?**	

```vue
<script>
  export function useModal() {
    const close = () => {}

    return { close }
  }
</script>

<script setup>
  function close() {

  }
</script>

<template>
  <Modal @close="close"></Modal>
</template>
```

**What did you expect to happen?**
No errors reported for the **no-shadow** rule.

**What actually happened?**

```
[3:11]: 'close' is already declared in the upper scope on line 10 column 12. (no-shadow)
```

**Repository to reproduce this issue**

[Online Playground](https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJyNV8lu5DYQ/RWhTwkwhIEcB7aRH8gf+EJRJTXHFKkhqV5i+N/zimot1OL40i29KhZrr9LHSbmKTj9Pz0F53cXXN1sUdOucj0XdWxW1s0Uf6B9XSfPHn8UH04tCORtioYwLVLwUwF9ei4/PNztQPcXe2+LjwfDJKH6en6Y78DI8FwG8Xbp1ui6dGu7aPRmp7YyMlE49J82Kv9Ohl7dT+n87vT4/JcIrH14cOP04+d5QOP38OF16elKubclGUWlPuP0CV/z1I1F+hZuA4UFcpA8T2vYmanF1vhI42jnLh61saWaxTkjv3VWMBgWhrbjKqM4ZT7hbxRSW00eqROddRz7qXNYOmU9VMsolW0UdLJDMyCThyl8wCLAy0ktW44ibQvTujieja1J3ZeiI0xnIAluIZMkHITv9NWtQULkKxsWvmOkCJ37FUGsTyR8RH16WRoxxPuA8x9YIMpQCrsMBl7ZGW/o/WRyMAKCWSAgRz4fivEPw2Gn2XUTZpIMHrMlbQsboddkf3sze/B7T9+RdRKktTiAbDzkOrbsIZ5H+XDiidZWu9WGcBta+LeGNkfVQbE/Ie1vrRrzTnTtUztnDLhC24AXxDSR0vaIYzYJnX2Qnh3bHRTV0GZF60pJjqgwha46mvEodlwxtH+EEO4Q3E96hezCB0BIy13iq0QME17S0VU4J5C9ww9rEibC5JWk8JSUKuldowVnIw1l6WnatdQcJuiJBNSyN4Rt9aSwQ1jLHbxE3SfgkRKnOmRK9RUtdKBF2iFm/ZThwXHUtrjqe8VTnfgQZ/TzM+uwHOZ3LlOZ8VGdtMt+nNr0bZU+/ewyJhQMXVTES2VUi3ruUvkMYFtqOXJ5sBfnDmNxQH+0yzPNjwxJd0yAZtU0xi0igoLMWPzIOVi8jNFGk0dXUwLKuNOi1lwL3HR5qdQyDOLk0dllLO4N0iCpCQDepZi9PanEPTqL3SXkFDCSLzBNRq/cVPgXdO7e+aeh+GxBbhFyL2faWJb4Bl2k6YjyBNuCOQL027SJaat0WRGfcClzkwQyprYad32Lh7K5b0Oz4jes8gYtMFwFNeDCb8+6xVD30aSjGPOunLQjRbDV8SzdS/ZhEvPtIgIsVJkwEIwPXSNDNJIqTFY0HadCIf8m7Gcd4W7NC5T1MoqyZf66mgYINyUDzhm4jWFHZNw35+d1Q5LKaER5I0jchAwbNEc1yHH8jZZldmyGXjTF2y4gitx9lOb6gqcFdin2dLsuJneQoTLbRbeUGRNVLUTpnSFq+KeaUgKIckVoaE8/YcJrzBKFnrSQ2xpVYznJQt2nmrkCLtXI33toOaZdC0E2ox3vPi+b1rNHvO6kmv+AzAMO15iRXOiyiiUSDaRXP5ANHtfrGKxRLQ44heaIsJ6KlK3altnRmQpx1KsJCKK5b/FNQspsUwSIONxozSUjM4wsyP7o0MMoeW66eTcawTo6gGYDtD7VGDLOvXjkx7JUZhn+FLxLPq4FW3NHnbxYm85bCyerlnG+81IqSUMi0qOs0p9GE5xcEhJLI9G3Ey/NMwx4AJ5dmAQVZEzZ67OvmvkJRuXJZeA/YdY/9HqK0ReBmeloYjCxpdu96i1isCKVU71i7CPN3TpSRqKYPtAWYh5IXkPQ8DtG7psf6wONMBysfrXeYOwirq/MQ4O3z8z8EAWCC)

**Comment by FloEdelmann** [comment]
Another repository to reproduce this issue was provided in #2141: https://github.com/teleskop150750/vue-no-shadow
